### DISCUSSION TITLE
SQS first send message command 20x/30x slower than following ones

### DISCUSSION THREAD

**Discussion by emdotem** [aws/aws-sdk-php#2427] [Q&A] [answered]
- [x] I've gone though [Developer Guide](https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/welcome.html) and [API reference](https://docs.aws.amazon.com/aws-sdk-php/v3/api/index.html)
- [x] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-php-sdk) for answers

**Version of AWS SDK for PHP?**
v3.179.2

**Version of PHP (`php -v`)?**
7.4.2

**Describe the question**
I'm using SQS to send heavy jobs to the queue when hitting certain APIs. I noticed that the API latency is around 1100ms since I introduced SQS. 

By running extra tests, I noticed that the latency of the first SQS message is 20x/30x slower than the ones that follow. 
What could be the reason?

In the attached image I created a script (which ran on an EC2 instance with IAM role enable to push message to the queue) which sends 800 messages to the queue, simple for loop with `$sqs->sendMessage` call.

<img width="927" alt="Screenshot 2021-08-14 at 00 48 57" src="https://user-images.githubusercontent.com/5315560/129425139-1faf87e8-81a9-4126-8408-2664dcab1936.png">

What I noticed it's an incredible high latency for the first SQS message. Every row it's an iteration of the for loop, hence a new message. You can also see the Partial time and the total time.

Any idea?

**Comment by yenfryherrerafeliz** [comment, marked as answer]
Hi @emdotem, since I already answered this issue [here](https://github.com/aws/aws-sdk-php/issues/2506) I will close this one. Basically, increasing the hop limit for ec2 instance from 1 to >1 should fix the issue. Basically, due that there hop limit defined for the ec2 instance metadata service is one, it will work fine from the ec2 instance, but if using docker, when resolving the credentials using the instance profile provider, the request to get the token from there will just time out ,with not responses. Then the secureMode variable is set to false [here](https://github.com/aws/aws-sdk-php/blob/69e83cc7947f545fa6e3b20d97403d750b72066f/src/Credentials/InstanceProfileProvider.php#L98), which makes the credentials to be retrieved without the token option and the explains why we get valid credentials. So, the latency that you are experiencing is due that the request being made to retrieve the token is not going anywhere and the implementation waits for a timeout then.

[Here](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InstanceMetadataOptionsResponse.html) you can find information about  httpPutResponseHopLimit.

Thanks!

[a comment by an automated account is not shown]
